src/bindings/python/flux/uri/resolvers/lsf.py

```diff
@@ -0,0 +1,84 @@
+###############################################################
+# Copyright 2022 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+###############################################################
+
+"""
+This module supports the Flux URI resolver for IBM Spectrum LSF.
+
+It can be used like:
+  `flux uri --remote lsf:<JOBID>`.
+
+"""
+
+import os
+import re
+import subprocess
+from pathlib import PurePath
+
+import yaml
+
+from flux.uri import URIResolverPlugin
+
+
+def lsf_find_compute_node(jobid):
+    """Figure out where the job is being run using JSON output from `bjobs -l`."""
+
+    sp = subprocess.run(
+        ["/opt/ibm/csm/bin/csm_allocation_query", "-j", str(jobid)], capture_output=True
+    )
+    try:
+        return str(yaml.safe_load(sp.stdout.decode("utf-8"))["compute_nodes"][0])
+    except:
+        raise ValueError("Unable to find a compute node attached to your job.")
+
+
+def lsf_job_pid(jobid, hostname):
+    """Resolve the pids of an LSF job using ssh to the compute node"""
+    sp = subprocess.run(
+        ["ssh", hostname, "ps", "-Ho", "pid,comm", "-u", os.getlogin()],
+        capture_output=True,
+    )
+    if sp.returncode != 0:
+        raise ValueError(f"Unable to find a flux session on host {hostname}")
+    for line in sp.stdout.decode("utf-8").split("\n"):
+        if "flux-broker" in line:
+            return line.split()[0]
+    raise ValueError(f"Unable to find a flux session on host {hostname}")
+
+
+def lsf_get_uri(hostname, pid):
+    """
+    Resolve the URI of a Flux instance by running
+
+      `ssh <HOSTNAME> flux uri --remote pid:<PID>`
+
+    and converting the result from bytecode to a string.
+    """
+    sp = subprocess.run(
+        ["ssh", hostname, "flux", "uri", "--remote", f"pid:{pid}"], capture_output=True
```

Comment:
I like this idea. I think it's a bit confusing though so I wish it were possible to relaunch the plugin with some additional command-line option or something to say that it's on the target host and so can find the PID and use the PID resolver (rather than doing a hostname check), but it doesn't look like that's possible given the way the URI resolver works.

The clearest thing I think would be to have some external script that would do nothing but find PIDs and URIs in one go. But adding a standalone script just for that would be overkill probably.

Reply:
You can pass options to the resolver by using URI `query` parameters as documented in [flux.uri.uri.FluxURIResolver.resolve](https://flux-framework.readthedocs.io/projects/flux-core/en/latest/python/autogenerated/flux.uri.uri.html#flux.uri.uri.FluxURIResolver.resolve). The query parameters should be available from the underlying [flux.uri.uri.URI](https://flux-framework.readthedocs.io/projects/flux-core/en/latest/python/autogenerated/flux.uri.uri.html#flux.uri.uri.URI) `query_dict` dictionary.

After saying all that though, I wonder if the least confusing way to do this would be to add a separate `first-local-broker` resolver or something, which would return the URI of the first broker found on the node for the given user. Then the steps used by the `lsf` resolver would be:

 1. determine first compute node of allocation
 2. run `ssh <host> flux uri resolve first-local-broker:` (The `:` is required to hint this is a URI not a jobid)

Either approach would be fine with me.